api/src/controllers/AuthenticationController.ts

```diff
@@ -83,11 +99,82 @@ export default class AuthenticationContoller extends ErrorController {
         throw ErrorController.NotFoundError("User not found");
       }
       const { scope } = userAttributes;
-      const accessToken = this.authService.generateAccessToken(scope);
+      const accessToken = this.authService.generateAccessToken(scope, user.username);
       Logger.info(`Successfully refreshed token for: ${user.username}`);
       res.status(200).send(accessToken).end();
     } catch (err) {
       next(err);
     }
   }
+
+  public async initMfa(_: Request, res: Response, next: NextFunction) {
+    try {
+      const user: TokenPayload | undefined = res.locals.user;
+      if (user === undefined) {
+        Logger.error("Missing user locals within initMFa controller.");
+        throw ErrorController.InternalServerError();
+      }
+      if (user.token_use !== TokenUse.Access) {
+        throw ErrorController.ForbiddenError("Unexpected token type.");
+      }
+      const userDetails: UserStoredAttributes | null = await this.userService.getUser(user.sub);
+      if (userDetails === null) {
+        throw ErrorController.ForbiddenError();
+      }
+      if (userDetails.mfaSecret !== undefined) {
+        Logger.error(`User ${user.sub} attempted to override MFA secret.`);
+        throw ErrorController.ForbiddenError();
+      }
+
+      const redis = getRedisClient();
+      const secret = this.authService.generateSecret(20);
+      await redis.set(user.sub, secret);
```

Comment:
I guess utilising the DB will be in the next PR?

Reply:
This is part of `initMfa`, where it is stored in the `redis` cache. The secret is written to the database for the user within `enableMfa`. This is so that the user is required to provide back a correct code, to ensure that they have saved the secret correctly and are not locked out in the event of an error.